clocky - an example of binding an atom

dribnet opened the pull request:
Little toy radial clock serves as an example of c2 bind!/unify with atoms and how to animate by resetting the atom. Note that this depends on an included javascript polyfill for requestAnimationFrame and currently doesn't render correctly under firefox (works well in webkit, including mobile safari and chrome).

dribnet committed:
clocky: a silly example to show svg animation

A primitive clock in 47 lines of clojurescript.

Note that the animation loop is only resetting an
atom based on the current time. The redrawing
logic is happening courtesy of c2 bindings.

dribnet committed:
clocky: updated css with simpler colors

Reduced the number of colors to two. Perhaps not the
right two colors, but definitely an improvement. This
works because of the stroke outline.

dribnet committed:
html: Fixed poorly formatted head

lynaghk commented on clocky/src/cljs/clocky/core.cljs:
```diff
@@ -0,0 +1,47 @@
+
+(defn nextloop []
+  (let [d (js/Date.)
+        h {}
```
This map construction seems very non-idiomatic.
Why not just use a literal?

``` clojure
(reset! clock-atom
  {"hours" ...
   "minutes" ...})
```

dribnet replied:
Thanks Kevin, I'm not sure what I was thinking there when I wrote that a week or two ago.

I cleaned this up with a couple of followup commits, including switching the map keys to keywords which I think looks a little cleaner. Note that I also replaced the map with an array-map since the order that these elements are drawn in is important (this assumes that unify will iterate through the map in order).

wemotom committed:
clocky: Cleaned up clock-atom reset

Replaced my strange assoc-let-chain with more idomatic
map literal.

wemotom committed:
clocky: clock-atom now array-map with keyword keys

The order of the clock-atom is important because the
clock needs to be rendered with the largest arcs
first, so hours, then minutes, etc. So we should use
an array-map to hold the contents of this data instead
of just a map. Note that this makes some light assumptions
on the inner workings of unify.

Also replaced the string keys with keywords for slightly
cleaner looking code, at the expense of calling (name)
a couple of times.

wemotom committed:
clocky: unify replaced with for, clock-atom now map

Since order of processing is important, replaced unify
with a for comprehension over the 4 keys. This means
we can also drop the array-map and use a simple map.

Note however that this version no longer works because
the bind! macro no longer seems to be watching the
clock-atom when it is not part of a unify. I'll
followup with a separate commit that fixes this.

lynaghk commented:
I believe the issue is because the `for` comprehension is lazy, and so the dereferencing of the atom isn't getting picked up.
Can you try dereferencing it once within a `let` outside of the `for` comprehension?

If that turns out what is going on, I'll update the underlying `reflex` library code to force evaluation of the body and push out a new release.

wemotom committed:
clocky: Fixed bind! failure by forcing side effects with doall

The for comprehension is lazy and this prevents the c2 atom
rebindings from working. Wrapping in doall provides the
desired behavior.

lynaghk commented:
Cool, thanks. Merging. I'll add a fix for this laziness issue in the next c2/reflex release.

dribnet commented:
Hi -  yep I confirmed this by wrapping the for in a doall.
